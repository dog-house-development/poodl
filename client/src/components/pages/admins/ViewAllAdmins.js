import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import _ from 'lodash';
import { Link, withRouter } from 'react-router-dom';

import AdminActions from '../../../redux/actions/adminActions';
import DataGrid from '../../ui/DataGrid';

export class ViewAllAdmins extends Component {
    constructor(props) {
        super(props);
        this.handleRowClick = this.handleRowClick.bind(this);
    }

    componentDidMount() {
        // call redux action to retrieve all admins from api
        this.props.adminActions.filter();
    }

    handleRowClick(e, id) {
        e.preventDefault();
        this.props.history.push(`/admins/${id}`);
    }

    getSuperColumnData(admin) {
        if (this.props.adminIsSuper) {
            return { super: admin.accessLevel === 'Super' ? 'Yes' : 'No' };
        }
    }

    getDataGridContent() {
        // choose what we want to display out of the admins data
        let data = [];
        _.each(this.props.admins, admin => {
            // we want admins' names and emails, and we need a key which will not be displayed
            data.push({
                firstName: admin.firstName,
                lastName: admin.lastName,
                email: admin.email,
                ...this.getSuperColumnData(admin),
                key: admin._id
            });
        });
        return data;
    }

    render() {
        return (
            <div className="page-container">
                <Link to="/dashboard" className="button small tertiary icon">
                    <i className="material-icons button-icon">keyboard_backspace</i> Back to home
                </Link>
                <div className="page-header">
                    <h1>Admins</h1>
                    <div className="button-list">
                        <Link to="/admins/register" className="button small primary icon">
                            <i className="material-icons button-icon">person_add</i>
                            Add admin
                        </Link>
                    </div>
                </div>
                <DataGrid
                    data={this.getDataGridContent()}
                    loading={this.props.loading}
                    onRowClick={this.handleRowClick}
                    noDataMessage="No admins to display"
                />
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        admins: state.admins.all,
        loading: state.admins.loading,
        errors: state.admins.errors,
        adminIsSuper: state.auth.admin.accessLevel === 'Super'
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        adminActions: bindActionCreators(AdminActions, dispatch)
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(ViewAllAdmins));
