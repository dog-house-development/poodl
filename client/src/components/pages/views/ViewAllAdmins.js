import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { Link } from 'react-router-dom';

import { fetchAdmins } from '../../../actions/adminActions';
import DataGrid from '../../ui/DataGrid';

export class ViewAllAdmins extends Component {
    componentDidMount() {
        // call redux action to retrieve all admins from api
        this.props.getAdmins();
    }

    getDataGridContent() {
        // choose what we want to display out of the admins data
        let data = [];
        _.each(this.props.admins, admin => {
            // we want admins' names and emails, and we need a key which will not be displayed
            data.push({
                firstName: admin.firstName,
                lastName: admin.lastName,
                email: admin.email,
                super: admin.superAdmin ? 'Yes' : 'No',
                key: admin._id,
                viewProfile: (
                    <Link to={'/admin/' + admin._id} className="button medium primary">
                        View
                    </Link>
                )
            });
        });
        return data;
    }

    render() {
        return (
            <div className="view-all-container">
                <Link to="/dashboard" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to home
                </Link>
                <h1>View All Admins</h1>
                <DataGrid data={this.getDataGridContent()} loading={this.props.loading} />
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        admins: state.admins.all,
        loading: state.admins.loading,
        errors: state.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        getAdmins: () => dispatch(fetchAdmins())
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ViewAllAdmins);
