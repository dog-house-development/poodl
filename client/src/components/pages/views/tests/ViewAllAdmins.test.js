import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';

import { Link } from 'react-router-dom';
import { ViewAllAdmins, mapStateToProps, mapDispatchToProps } from '../ViewAllAdmins';

configure({ adapter: new Adapter() });

describe('ViewAllAdmins tests', () => {
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                admins: {
                    loading: false,
                    all: [
                        {
                            _id: '123',
                            firstName: 'Big',
                            lastName: 'Tup',
                            email: 'bigtup@nowhere.com',
                            superAdmin: true
                        },
                        {
                            _id: '321',
                            firstName: 'Lil',
                            lastName: 'Tup',
                            email: 'liltup@nowhere.com',
                            superAdmin: false
                        }
                    ]
                },
                errors: {}
            },
            _state
        );
        props = _.assign({}, _props);
        wrapper = shallow(
            <ViewAllAdmins
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('mapStateToProps', () => {
        it('should map state to props', () => {
            expect(mapStateToProps(state, props)).toEqual({
                admins: [
                    {
                        _id: '123',
                        firstName: 'Big',
                        lastName: 'Tup',
                        email: 'bigtup@nowhere.com',
                        superAdmin: true
                    },
                    {
                        _id: '321',
                        firstName: 'Lil',
                        lastName: 'Tup',
                        email: 'liltup@nowhere.com',
                        superAdmin: false
                    }
                ],
                errors: {},
                loading: false
            });
        });
    });

    describe('mapDispatchToProps', () => {
        it('should map dispatch to props', () => {
            const dispatch = jest.fn();
            expect(JSON.stringify(mapDispatchToProps(dispatch))).toEqual(JSON.stringify({ getAdmins: () => {} }));
        });
    });

    describe('componentDidMount', () => {
        it('should run without breaking', () => {
            spyOn(instance, 'componentDidMount');
            instance.componentDidMount();
        });
    });

    describe('getDataGridContent', () => {
        it('should return filtered admins data', () => {
            expect(instance.getDataGridContent()).toEqual([
                {
                    key: '123',
                    firstName: 'Big',
                    lastName: 'Tup',
                    email: 'bigtup@nowhere.com',
                    super: 'Yes',
                    viewProfile: (
                        <Link to={'/admin/123'} className="button medium primary">
                            View
                        </Link>
                    )
                },
                {
                    key: '321',
                    firstName: 'Lil',
                    lastName: 'Tup',
                    email: 'liltup@nowhere.com',
                    super: 'No',
                    viewProfile: (
                        <Link to={'/admin/321'} className="button medium primary">
                            View
                        </Link>
                    )
                }
            ]);
        });
    });

    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
