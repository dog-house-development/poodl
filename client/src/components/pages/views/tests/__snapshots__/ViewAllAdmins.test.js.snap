// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewAllAdmins tests render should render correctly 1`] = `
<div
  className="view-all-container"
>
  <Link
    className="button small tertiary"
    replace={false}
    to="/dashboard"
  >
    <i
      className="material-icons"
    >
      keyboard_backspace
    </i>
     Back to home
  </Link>
  <h1>
    View All Admins
  </h1>
  <DataGrid
    data={
      Array [
        Object {
          "email": "bigtup@nowhere.com",
          "firstName": "Big",
          "key": "123",
          "lastName": "Tup",
          "super": "Yes",
          "viewProfile": <Link
            className="button medium primary"
            replace={false}
            to="/admin/123"
          >
            View
          </Link>,
        },
        Object {
          "email": "liltup@nowhere.com",
          "firstName": "Lil",
          "key": "321",
          "lastName": "Tup",
          "super": "No",
          "viewProfile": <Link
            className="button medium primary"
            replace={false}
            to="/admin/321"
          >
            View
          </Link>,
        },
      ]
    }
    loading={false}
  />
</div>
`;
